File: app/src/main/java/uk/co/sundroid/dao/DatabaseHelper.java
package uk.co.sundroid.dao;

import java.util.ArrayList;

import uk.co.sundroid.domain.LocationDetails;
import uk.co.sundroid.util.geo.LatitudeLongitude;
import uk.co.sundroid.util.time.TimeZoneResolver;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

public class DatabaseHelper extends SQLiteOpenHelper {
	
	private static final String DATABASE_NAME = "sundroid";
	
	private static final String LAT = "lat";
	private static final String LON = "lon";
	private static final String NAME = "name";
	private static final String COUNTRY = "country";
	private static final String TIMEZONE_ID = "zoneId";
	private static final String TIMESTAMP = "tstamp";

	public DatabaseHelper(Context context) {
		super(context, DATABASE_NAME, null, 6);
	}
	
	@Override
	public void onCreate(SQLiteDatabase db) {
		db.execSQL("CREATE TABLE saved_locations (id INTEGER PRIMARY KEY AUTOINCREMENT, " +
					"lat REAL, lon REAL, name TEXT, country TEXT, zoneId TEXT, zoneOffset TEXT, tstamp INTEGER);");
	}

	@Override
	public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

	}
	
	public void addSavedLocation(LocationDetails location) {
		SQLiteDatabase db = getWritableDatabase();
		ContentValues cv = new ContentValues();
		cv.put(LAT, location.getLocation().getLatitude().getDoubleValue());
		cv.put(LON, location.getLocation().getLongitude().getDoubleValue());
		cv.put(NAME, location.getName());
		cv.put(COUNTRY, location.getCountry());
		cv.put(TIMEZONE_ID, location.getTimeZone().getId());
		cv.put(TIMESTAMP, System.currentTimeMillis());
		db.insert("saved_locations", NAME, cv);
	}

	public ArrayList<LocationDetails> getSavedLocations() {
		SQLiteDatabase db = getReadableDatabase();
		Cursor cursor = db.rawQuery("SELECT id, lat, lon, name, country, zoneId FROM saved_locations ORDER BY name ASC", null);
		
		cursor.moveToFirst();
		ArrayList<LocationDetails> savedLocations = new ArrayList<>();
		while (!cursor.isAfterLast()) {
			LocationDetails savedLocation = new LocationDetails();
			savedLocation.setId(cursor.getInt(0));
			savedLocation.setLocation(new LatitudeLongitude(cursor.getDouble(1), cursor.getDouble(2)));
			savedLocation.setName(cursor.getString(3));
			savedLocation.setCountry(cursor.getString(4));
			savedLocation.setTimeZone(TimeZoneResolver.getTimeZone(cursor.getString(5), true));
			savedLocations.add(savedLocation);
			cursor.moveToNext();
		}
		cursor.close();
		return savedLocations;
	}

    public LocationDetails getSavedLocation(int id) {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT id, lat, lon, name, country, zoneId FROM saved_locations WHERE id = " + id, null);

        LocationDetails savedLocation = null;
        cursor.moveToFirst();
        if (!cursor.isAfterLast()) {
            savedLocation = new LocationDetails();
            savedLocation.setId(cursor.getInt(0));
            savedLocation.setLocation(new LatitudeLongitude(cursor.getDouble(1), cursor.getDouble(2)));
            savedLocation.setName(cursor.getString(3));
            savedLocation.setCountry(cursor.getString(4));
            savedLocation.setTimeZone(TimeZoneResolver.getTimeZone(cursor.getString(5), true));
        }
        cursor.close();
        return savedLocation;
    }
	
	public void deleteSavedLocation(long rowId) {
		SQLiteDatabase db = getReadableDatabase();
		String[] args = { String.valueOf(rowId) };
		db.delete("saved_locations", "id = ?", args);
	}
	
}

File: app/src/main/kotlin/uk/co/sundroid/util/dao/DatabaseHelper.kt
package uk.co.sundroid.util.dao

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import uk.co.sundroid.domain.LocationDetails
import uk.co.sundroid.util.location.LatitudeLongitude
import uk.co.sundroid.util.time.TimeZoneResolver
import java.util.*

class DatabaseHelper(context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, 7), AutoCloseable {

    val savedLocations: ArrayList<LocationDetails>
        get() {
            val db = readableDatabase
            val cursor = db.rawQuery("SELECT id, lat, lon, name, country, zoneId FROM saved_locations ORDER BY name ASC", null)

            cursor.moveToFirst()
            val savedLocations = ArrayList<LocationDetails>()
            while (!cursor.isAfterLast) {
                val savedLocation = LocationDetails(LatitudeLongitude(cursor.getDouble(1), cursor.getDouble(2)))
                savedLocation.id = cursor.getLong(0)
                savedLocation.name = cursor.getString(3)
                savedLocation.country = cursor.getString(4)
                savedLocation.timeZone = TimeZoneResolver.getTimeZone(cursor.getString(5))
                savedLocations.add(savedLocation)
                cursor.moveToNext()
            }
            cursor.close()
            return savedLocations
        }

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL("CREATE TABLE saved_locations (id INTEGER PRIMARY KEY AUTOINCREMENT, lat REAL, lon REAL, name TEXT, country TEXT, zoneId TEXT, zoneOffset TEXT, tstamp INTEGER);")
        db.execSQL("CREATE TABLE widget_locations (id INTEGER, lat REAL, lon REAL, name TEXT, country TEXT, zoneId TEXT, zoneOffset TEXT, tstamp INTEGER);")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (oldVersion < 7) {
            db.execSQL("CREATE TABLE widget_locations (id INTEGER, lat REAL, lon REAL, name TEXT, country TEXT, zoneId TEXT, zoneOffset TEXT, tstamp INTEGER);")
        }
    }

    fun addSavedLocation(location: LocationDetails) {
        val db = writableDatabase
        val cv = ContentValues()
        cv.put(LAT, location.location.latitude.doubleValue)
        cv.put(LON, location.location.longitude.doubleValue)
        cv.put(NAME, location.name)
        cv.put(COUNTRY, location.country)
        cv.put(TIMEZONE_ID, location.timeZone!!.id)
        cv.put(TIMESTAMP, System.currentTimeMillis())
        val id = db.insert("saved_locations", NAME, cv)
        location.id = id
    }

    fun deleteSavedLocation(rowId: Long) {
        val db = readableDatabase
        db.delete("saved_locations", "id = ?", arrayOf(rowId.toString()))
    }

    fun setWidgetLocation(widgetId: Int, location: LocationDetails) {
        deleteWidgetLocation(widgetId)
        val db = writableDatabase
        val cv = ContentValues()
        cv.put(ID, widgetId)
        cv.put(LAT, location.location.latitude.doubleValue)
        cv.put(LON, location.location.longitude.doubleValue)
        cv.put(NAME, location.name)
        cv.put(COUNTRY, location.country)
        cv.put(TIMEZONE_ID, location.timeZone!!.id)
        cv.put(TIMESTAMP, System.currentTimeMillis())
        db.insert("widget_locations", NAME, cv)
    }

    fun getWidgetLocation(widgetId: Int): LocationDetails? {
        val db = readableDatabase
        val cursor = db.rawQuery("SELECT id, lat, lon, name, country, zoneId FROM widget_locations WHERE id = $widgetId ORDER BY tstamp DESC", null)
        cursor.moveToFirst()
        var widgetLocation: LocationDetails? = null
        if (!cursor.isAfterLast) {
            widgetLocation = LocationDetails(LatitudeLongitude(cursor.getDouble(1), cursor.getDouble(2)))
            widgetLocation.id = cursor.getInt(0).toLong()
            widgetLocation.name = cursor.getString(3)
            widgetLocation.country = cursor.getString(4)
            widgetLocation.timeZone = TimeZoneResolver.getTimeZone(cursor.getString(5))
        }
        cursor.close()
        return widgetLocation
    }

    fun deleteWidgetLocation(widgetId: Int) {
        val db = readableDatabase
        val args = arrayOf(widgetId.toString())
        db.delete("widget_locations", "id = ?", args)
    }

    companion object {
        private const val DATABASE_NAME = "sundroid"
        private const val ID = "id"
        private const val LAT = "lat"
        private const val LON = "lon"
        private const val NAME = "name"
        private const val COUNTRY = "country"
        private const val TIMEZONE_ID = "zoneId"
        private const val TIMESTAMP = "tstamp"
    }

}
